File: app/src/main/java/com/talk51/sentenceparser/SentenceParser.java
package com.talk51.sentenceparser;

import android.util.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class SentenceParser {
    private final static String TAG = "SentenceParser";
    public SentenceParser() {

    }

    public static String parse(String sentence) {
        if (sentence.contains("+")) {
            // wrong format, don't parse
            return sentence;
        }

//        log(sentence);

        // replace "\t, \r, \n, ^, chinese( [^\x20-\x7e]* )" to null
        Pattern pattern = Pattern.compile("\\\\t|\\\\r|\\\\n|\\^|[^\\x20-\\x7e]*");
        Matcher matcher = pattern.matcher(sentence);
        sentence = matcher.replaceAll("");

//        log(sentence);

        // replace "-, (, )" to space
        pattern = Pattern.compile("[-()]");
        matcher = pattern.matcher(sentence);
        sentence = matcher.replaceAll(" ");

//        log(sentence);

        /* process pattern "A a" return "A" */
        String[] words = sentence.split(" ");
        if (words.length == 2 && words[0].equalsIgnoreCase(words[1])) {
            return words[0].toUpperCase();
        }

        /* process pattern {a # b $ c} */
        int choiceSize = checkChoiceSentence(sentence);
        if (choiceSize > 0) {
//            log("choiceSize = " + choiceSize);
            sentence = parseChoiceSentence(sentence);
        }

        // replace multi spaces to one space
        pattern = Pattern.compile("[ ]+");
        matcher = pattern.matcher(sentence);
        sentence = matcher.replaceAll(" ");

        return sentence;
    }

    /** Example: A {A # a} B {B # b $ bb $ bbb} C {C # c $ cc} D. */
    private static String parseChoiceSentence(String sentence) {
        StringBuilder builder = new StringBuilder();
        ArrayList<StringBuilder> sentences = new ArrayList<>();
        ArrayList<String> originList = new ArrayList<>();
        ArrayList<String> choiceList = new ArrayList<>();
        HashMap<String, ArrayList<String>> choiceSentences = new HashMap<>();

        String tempStr = sentence;
        int openIndex, closeIndex;
        while ((openIndex = tempStr.indexOf("{")) != -1) {
            originList.add(tempStr.substring(0, openIndex));

            closeIndex = tempStr.indexOf("}");
            String choice = tempStr.substring(openIndex + 1, closeIndex);
            if (choice.contains("#")) {
                String[] words = choice.split("#");
                if ((words.length == 2) &&
                    (words[0].length() > 0) &&
                    (words[1].length() > 0)) {
                    String key = words[0];
                    choiceList.add(key);
                    ArrayList<String> choices = new ArrayList<>(Arrays.asList(words[1].split("\\$")));
                    choiceSentences.put(key, choices);
                }
            }

            tempStr = tempStr.substring(closeIndex + 1);
        }
        // last partial if have
        if (tempStr.length() > 0) {
            originList.add(tempStr);
        }

        /* Combine all possibilities sentence */
        for (int i = 0; i < originList.size(); ++i) {
            if (i == 0) {
                StringBuilder sent = new StringBuilder();
                sent.append(originList.get(i));
                sentences.add(sent); // one element
            } else {
                for (StringBuilder sent : sentences) {
                    sent.append(originList.get(i));
                }
            }

            if (i < choiceList.size()) {
                String choice = choiceList.get(i);
                ArrayList<String> choices = choiceSentences.get(choice);
                int size = choices.size();
                ArrayList<StringBuilder> tmpSents = new ArrayList<>(sentences);
                sentences.clear();
                for (int j = 0; j < size; ++j) {
                    for (StringBuilder sent : tmpSents) {
                        StringBuilder tmpSB = new StringBuilder(sent);
                        tmpSB.append(choices.get(j));
                        sentences.add(tmpSB);
                    }
                }
            }
        }

        /* Combine all sentences in one */
        for (int idx = 0; idx < sentences.size(); ++idx) {
            if (idx != 0) {
                builder.append("|");
            }
            builder.append(sentences.get(idx));
//            log(sentences.get(idx).toString());
        }

        return builder.toString();
    }

    private static int checkChoiceSentence(String sentence) {
        int openCount = 0;
        int closeCount = 0;
        String tempStr = sentence;
        int openIndex, closeIndex;

        while ((openIndex = tempStr.indexOf("{")) != -1) {
            ++openCount;

            if ((closeIndex = tempStr.indexOf("}")) > openIndex) {
                ++closeCount;
            } else {
                break;
            }
            tempStr = tempStr.substring(closeIndex + 1);
        }

        if (openCount != closeCount) {
            return -1;
        }

        return openCount;
    }

    private static void log(String msg) {
        Log.i(TAG, msg);
    }
}

File: SentenceParser/SentenceParser.m
//
//  SentenceParser.m
//  SentenceParser
//
//  Created by VintonLiu 2019/8/4.
//  Copyright © 2019 VintonLiu. All rights reserved.
//

#import "SentenceParser.h"

@implementation SentenceParser
+ (NSString*)parse:(NSString*)sentence {
//    NSLog(@"%@", sentence);
//    if ([sentence containsString:@"+"]) {
//        return sentence;
//    }
    
    // replace "\t, \r, \n, ^, chinese( [^\x20-\x7e]* )" to null
    sentence = [sentence stringByReplacingOccurrencesOfString:@"\\^|\\\\t|\\\\r|\\\\n|[^\\x20-\\x7e]*"
                                                             withString:@""
                                                                options:NSRegularExpressionSearch
                                                                  range:NSMakeRange(0, sentence.length)];
//    NSLog(@"%@", sentence);
    
    // replace "-, (, )" to space
    sentence = [sentence stringByReplacingOccurrencesOfString:@"[-()]"
                                                   withString:@" "
                                                      options:NSRegularExpressionSearch
                                                        range:NSMakeRange(0, sentence.length)];
//    NSLog(@"%@", sentence);
    
    /* process pattern "A a" return "A" */
    NSArray* words = [sentence componentsSeparatedByString:@" "];
    if ((words.count == 2) &&
        ([[words objectAtIndex:0] caseInsensitiveCompare:[words objectAtIndex:1]] == NSOrderedSame)) {
        return [[words objectAtIndex:0] uppercaseString];
    }
    
    /* process pattern {a # b $ c} */
    int choiceSize = [SentenceParser checkChoiceSentence:sentence];
    if (choiceSize > 0) {
//        NSLog(@"choiceSize = %d", choiceSize);
        sentence = [SentenceParser parseChoiceSentence:sentence];
    }
    
    // replace multi spaces to one space
    sentence = [sentence stringByReplacingOccurrencesOfString:@"[  ]+"
                                                   withString:@" "
                                                      options:NSRegularExpressionSearch
                                                        range:NSMakeRange(0, sentence.length)];
//    NSLog(@"%@", sentence);
    return sentence;
}

+ (int) checkChoiceSentence: (NSString*) sentence {
    int openCount = 0;
    int closeCount = 0;
    NSString* tempStr = sentence;
    int openIndex = 0, closeIndex = 0;
    
    while ((openIndex = (int)[tempStr rangeOfString:@"{"].location) >= 0) {
        ++openCount;
        
        if ((closeIndex = (int)[tempStr rangeOfString:@"}"].location) > openIndex) {
            ++closeCount;
        } else {
            break;
        }
        
        tempStr = [tempStr substringFromIndex:(closeIndex + 1)];
    }
    
    if (openCount != closeCount) {
        return -1;
    }
    
    return openCount;
}

+ (NSString*) parseChoiceSentence: (NSString*) sentence {
    NSMutableArray *originList = [[NSMutableArray alloc] init];
    NSMutableArray *choiceList = [[NSMutableArray alloc] init];
    NSMutableDictionary *choiceSentences = [[NSMutableDictionary alloc] init];
    
    NSString* tempStr = sentence;
    int openIndex = 0, closeIndex = 0;
    while ((openIndex = (int)[tempStr rangeOfString:@"{"].location) >= 0) {
        closeIndex = (int)[tempStr rangeOfString:@"}"].location;
        [originList addObject:[tempStr substringToIndex:openIndex]];
        
        NSString *choice = [tempStr substringWithRange:NSMakeRange(openIndex + 1, (closeIndex - 1 - openIndex))];
        if ([choice containsString:@"#"]) {
            NSArray* words = [choice componentsSeparatedByString:@"#"];
            if ((words.count == 2) &&
                ([[words objectAtIndex:0] length] > 0) &&
                ([[words objectAtIndex:1] length] > 0)) {
                NSString *key = [words objectAtIndex:0];
                [choiceList addObject:key];
                
                NSArray * choices = [[words objectAtIndex:1] componentsSeparatedByString:@"$"];
                [choiceSentences setObject:choices forKey:key];
            }
        }
        
        tempStr = [tempStr substringFromIndex:closeIndex + 1];
    }
    // last partial if have
    if ([tempStr length] > 0) {
        [originList addObject:[tempStr substringFromIndex:0]];
    }
    
    /* Combine all possibilities sentence */
    NSMutableArray *sentences = [[NSMutableArray alloc] init];
    for (int i = 0; i < originList.count; ++i) {
        if (i == 0) {
            [sentences addObject:[originList objectAtIndex:i]];
        } else {
            for (int j = 0; j < sentences.count; ++j) {
                NSString* sent = [[sentences objectAtIndex:j] stringByAppendingString:[originList objectAtIndex:i]];
                [sentences replaceObjectAtIndex:j withObject:sent];
            }
        }
        
        if (i < choiceList.count) {
            NSString *key = [choiceList objectAtIndex:i];
            NSArray *choices = [choiceSentences objectForKey:key];
            
            NSMutableArray* tmpSents = [sentences mutableCopy];
            [sentences removeAllObjects];
            
            for (NSString* choice in choices) {
                for (NSString* sent in tmpSents) {
                    [sentences addObject:[sent stringByAppendingString:choice]];
                }
            }
        }
    }
    
    
    sentence = [sentences componentsJoinedByString:@"|"];
    
    return sentence;
}
@end
